### ISSUE TITLE
`<ranges>`: `elements_view::iterator::iterator_category` invokes the `operator*() &&` of the underlying iterator

### ISSUE THREAD

**Issue by hewillk** [microsoft/STL#3031]
https://github.com/microsoft/STL/blob/b812b0e22b6e2a777800c6080ab636d70aec0678/stl/inc/ranges#L4982-L4989

Should `_STD declval<iterator_t<_Base>>()` be `_STD declval<iterator_t<_Base>&>()`?

```cpp
#include <ranges>

struct I {
  using difference_type = std::ptrdiff_t;
  using value_type = std::pair<int, double>;
  value_type& operator*() const &;
  void operator*() && = delete;
  I& operator++();
  I operator++(int);
  bool operator==(const I&) const;
};

int main() {
  std::ranges::input_range auto r = std::ranges::subrange(I{}, std::unreachable_sentinel);
  auto e = r | std::views::keys;
}
```
https://godbolt.org/z/oW8dP56zv

**Comment by CaseyCarter** [comment]
The controlling spec is [range.elements.iterator]/2.1:

> If `std::get<N>(*current_)` is an rvalue, `iterator_category` denotes `input_iterator_tag`.

Where `current_` is an exposition-only non-static data member of type `iterator_t<Base>>`. That suggests that we want `declval<iterator_t<_Base>&>()`. Since the `iterator_t<_Base>` must model `input_iterator`, `decltype(*declval<iterator_t<_Base>&>())` and `decltype(*declval<const iterator_t<_Base>&>())` necessarily denote the same type. Given the equivalence, I think I'd prefer `*declval<const iterator_t<_Base>&>()` simply because it most closely reflects the `input_iterator` requirements.

**Comment by timsong-cpp** [comment]
Implicit expression variations in _`indirectly-readable-impl`_ apply and require those to be the same type.

**Comment by CaseyCarter** [comment]
> Implicit expression variations in _`indirectly-readable-impl`_ apply and require those to be the same type.

Yes, that's my "necessarily denote the same type". When implicit expression variations say there are several ways to spell an expression, we conventionally prefer to use the `const&` spelling that appears directly in the spec.

**Comment by timsong-cpp** [comment]
> > Implicit expression variations in _`indirectly-readable-impl`_ apply and require those to be the same type.
> 
> Yes, that's my "necessarily denote the same type". When implicit expression variations say there are several ways to spell an expression, we conventionally prefer to use the `const&` spelling that appears directly in the spec.

My point is that `iterator_t<_Base>` is also an acceptable spelling, and so there's no bug and the code in the OP is IFNDR. (I have no opinion on the best way to spell this.)
